src/models/db/case.js

```diff
@@ -1,72 +1,134 @@
+    .then((results) => {
+      if (results.length === 1) {
+        return results[0]
+      } else if (results.length === 0) {
+        return null
+      } else {
+        assert(false)
       }
-      data.notification_targets = JSON.parse(data.notification_targets)
-      delete data.created_at
-      delete data.updated_at
-      if (!Array.isArray(data.notaries)) {
-        data.notaries = [ data.notaries ]
+    })
+}
+
+function getCaseByUuid (caseUuid, options) {
+  return knexOrTransaction(knex, options)
+    .from(TABLE_NAME)
+    .select()
+    .where('CASE_UUID', caseUuid)
+    .then((results) => results.map(convertFromPersistent))
+    .then((results) => {
+      if (results.length === 1) {
+        return results[0]
+      } else if (results.length === 0) {
+        return null
+      } else {
+        assert(false)
       }
-      return data
-    }
+    })
+}
 
-    static convertToPersistent (data) {
-      data.execution_condition = JSON.stringify(data.execution_condition)
-      data.exec_cond_fulfillment = JSON.stringify(data.exec_cond_fulfillment)
-      data.notification_targets = JSON.stringify(data.notification_targets)
-      data.notaries = data.notaries[0]
-      return data
-    }
+function save (caseObject, options) {
+  const myKnex = knexOrTransaction(knex, options)
+  const persistentCaseObject = convertToPersistent(caseObject)
+  if (persistentCaseObject.CASE_ID) {
+    return myKnex(TABLE_NAME).update(persistentCaseObject)
+      .where('CASE_ID', persistentCaseObject.CASE_ID)
+      .then()
+  } else {
+    return myKnex(TABLE_NAME).insert(persistentCaseObject).then()
+  }
+}
 
-    isFinalized () {
-      return this.state === 'executed' || this.state === 'rejected'
-    }
+function * _upsertCase (caseData, options) {
+  assert(options.transaction)
+  // SQLite's implementation of upsert does not tell you whether it created the
+  // row or whether it already existed. Since we need to know to return the
+  // correct HTTP status code we unfortunately have to do this in two steps.
+  const existingCase = yield getCaseByUuid(caseData.case_uuid, options)
+  if (existingCase) {
+    yield save(_.merge({}, existingCase, caseData), options)
+  } else {
+    yield save(caseData, options)
   }
+  return Boolean(existingCase)
+}
 
-  Case.validateExternal = validator.create('Case')
-  Case.tableName = 'cases'
+function * upsertCase (caseData, options) {
+  if (options && options.transaction) {
+    return yield _upsertCase(caseData, options)
+  } else {
+    let result
+    yield transaction(function * (transaction) {
+      result = yield _upsertCase(caseData,
+        _.assign({}, options || {}, {transaction}))
+    })
+    return result
+  }
+}
 
-  PersistentKnexModelMixin(Case, knex)
+function convertFromPersistent (data) {
+  data = _.cloneDeep(data)
+  data = _.mapKeys(data, (value, key) => key.toLowerCase())
+  data.case_status_id = CASE_STATUS_IDS[data.case_status_id]
+  if (!Array.isArray(data.notaries)) {
```

Comment:
I'm not sure why the database stores the first notary in the array, while the JSON allows multiple notaries in the array. Looks like Nao added the code originally, and he's not around to ask. Before I change this, I'd like to check with @stefan to determine the use case for multiple notaries per case.